XMonad/Hooks/Rescreen.hs

```diff
@@ -0,0 +1,146 @@
+-- and adding 'rescreenHook' to your 'xmonad' config:
+--
+-- > main = xmonad $ … . rescreenHook rescreenCfg . … $ def{…}
+
+-- | Hook configuration for 'rescreenEventHook'.
+data RescreenConfig = RescreenConfig
+    { afterRescreenHook :: X () -- ^ hook to invoke after 'rescreen'
+    , randrChangeHook :: X () -- ^ hook for other randr changes, e.g. (dis)connects
+    }
+
+instance Default RescreenConfig where
+    def = RescreenConfig
+        { afterRescreenHook = mempty
+        , randrChangeHook = mempty
+        }
+
+-- | Attach custom hooks to screen (xrandr) configuration change events.
+-- Replaces the built-in rescreen handling of xmonad core with:
+--
+-- 1. listen to 'RRScreenChangeNotifyEvent' in addition to 'ConfigureEvent' on
+--    the root window
+-- 2. whenever such event is received:
+-- 3. clear any other similar events (Xorg server emits them in bunches)
+-- 4. if any event was 'ConfigureEvent', 'rescreen' and invoke 'afterRescreenHook'
+-- 5. if there was no 'ConfigureEvent', invoke 'randrChangeHook' only
+--
+-- 'afterRescreenHook' is useful for restarting/repositioning status bars and
+-- systray.
+--
+-- 'randrChangeHook' may be used to automatically trigger xrandr (or perhaps
+-- autorandr) when outputs are (dis)connected.
+rescreenHook :: RescreenConfig -> XConfig a -> XConfig a
+rescreenHook cfg xConfig =
+    xConfig{ startupHook = startupHook xConfig <> rescreenStartupHook
+           , handleEventHook = handleEventHook xConfig <> rescreenEventHook cfg
+           }
+
+-- | Startup hook to listen for @RRScreenChangeNotify@ events.
+rescreenStartupHook :: X ()
+rescreenStartupHook = do
+    dpy <- asks display
+    root <- asks theRoot
+    io $ xrrSelectInput dpy root rrScreenChangeNotifyMask
+
+-- | Event hook with custom rescreen/randr hooks. See 'rescreenHook' for more.
+rescreenEventHook :: RescreenConfig -> Event -> X All
+rescreenEventHook cfg e = do
+    shouldHandle <- case e of
+        ConfigureEvent{ ev_window = w } -> isRoot w
+        RRScreenChangeNotifyEvent{ ev_window = w } -> isRoot w
+        _ -> pure False
+    if shouldHandle
+        then All False <$ handleEvent cfg e
+        else mempty
+
+handleEvent :: RescreenConfig -> Event -> X ()
+handleEvent RescreenConfig{..} e = do
+    -- Xorg emits several events after every change, clear them to prevent
+    -- triggering the hook multiple times.
+    moreConfigureEvents <- clearTypedWindowEvents (ev_window e) configureNotify
+    _ <- clearTypedWindowRREvents (ev_window e) rrScreenChangeNotify
+    -- If there were any ConfigureEvents, this is an actual screen
+    -- configuration change, so rescreen and fire rescreenHook. Otherwise,
+    -- this is just a connect/disconnect, fire randrChangeHook.
+    if ev_event_type e == configureNotify || moreConfigureEvents
+        then rescreen >> afterRescreenHook
+        else randrChangeHook
+
+-- | Remove all X events of a given window and type from the event queue,
+-- return whether there were any.
```

Comment:
This looks really cool; not only in the context of this module but in general even.  Do we have a place where these kinds of "XUtils" (sadly `XMonad.Util.XUtils` does not seem to be this place) can go?

Reply:
Not currently. Sadly evicting and reclaiming `XMonad.Util.XUtils` would be a breaking change; but it's frankly misnamed.

That said, maybe it deserves to be in `XMonad.Util.EvUtils` (this doesn't exist yet).

Reply:
What's wrong with `XMonad.Util.XUtils`?

On the other hand: should I really try to move a function that nobody needed for a decade somewhere where it's only slightly easier to find than where it is? After all, if anyone opens a PR that adds a similar function for their module, surely I'll notice and suggest factoring it out then. :-)

Reply:
(Anyway, if we ever make the function available, it should come with some warnings as it may change the order of events. It does put them back in the correct order relative to one another, but the order is changed relative to events that weren't picked up by `XCheckTypedWindowEvent`. It is extremely unlikely to matter in this usecase, but with something else than `ConfigureEvent`, it may matter.)

Reply:
> What's wrong with `XMonad.Util.XUtils`?

The module description says "A module for painting on the screen" and that's pretty much what all of the functions do in some capacity; not really a place for general X utility functions.  Then again this is only there as a doc, so we may well repurpose it and give all the drawing stuff a heading of its own.

> On the other hand: should I really try to move a function that nobody needed for a decade somewhere where it's only slightly easier to find than where it is? After all, if anyone opens a PR that adds a similar function for their module, surely I'll notice and suggest factoring it out then. :-)

I'll give you that one; probably a bit of a premature optimization, especially since no good module already exists where we could put it.

Reply:
> The module description says "A module for painting on the screen" and that's pretty much what all of the functions do in some capacity; not really a place for general X utility functions.

There are also a couple of X wrappers for IO functions with Display parameter. We can move those under a separate haddock heading and then add another heading for utilities (which I define as more than a wrapper) like my clearTypedWindowEvents. None of that needs to be done now, but I think this is the most sensible thing to do if we want a module for X utils.